File: backtracking/permutations/permutations.go
package permutations

func permute(nums []int) [][]int {
	ans := [][]int{}
	list := []int{}
	visited := make([]bool, len(nums))

	backtrack(nums, &visited, &ans, &list)

	return ans
}

func backtrack(nums []int, visited *[]bool, ans *[][]int, list *[]int) {
	if len(*list) == len(nums) {
		listCopy := make([]int, len(*list))
		copy(listCopy, *list)

		*ans = append(*ans, listCopy)
	} else {
		for i := range nums {
			if (*visited)[i] {
				continue
			}

			(*visited)[i] = true
			*list = append(*list, nums[i])

			backtrack(nums, visited, ans, list)

			*list = (*list)[:len(*list)-1]
			(*visited)[i] = false
		}
	}
}

File: backtracking/permutations/permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
function permute(nums) {
  const ans = [];
  const list = [];
  const visited = new Set();

  backtrack(nums, visited, ans, list);

  return ans;
}

function backtrack(nums, visited, ans, list) {
  if (list.length == nums.length) {
    ans.push(list.slice());
    return;
  }

  for (let i = 0; i < nums.length; i++) {
    if (visited.has(i)) {
      continue;
    }

    visited.add(i);
    list.push(nums[i]);

    backtrack(nums, visited, ans, list);

    list.pop();
    visited.delete(i);
  }
}
